src/tracksdata/graph/_rustworkx_graph.py

```diff
@@ -1086,48 +1086,74 @@ def assign_track_ids(
+            else:
+                previous_id_df = self.node_attrs(attr_keys=[DEFAULT_ATTR_KEYS.NODE_ID, output_key])
+                if track_id_offset is None:
+                    existing_ids = previous_id_df[output_key].unique().to_list()
+                    existing_ids = [i for i in existing_ids if i != -1]
+                    track_id_offset: int = max(existing_ids, default=0) + 1
 
-        Returns
-        -------
-        rx.PyDiGraph
-            A compressed graph (parent -> child) with track ids lineage relationships.
-        """
-        try:
-            node_ids, track_ids, tracks_graph = _assign_track_ids(self.rx_graph, track_id_offset)
-        except RuntimeError as e:
-            raise RuntimeError(
-                "Are you sure this graph is a valid lineage graph?\n"
-                "This function expects a solved graph.\n"
-                "Often used from `graph.subgraph(edge_attr_filter={'solution': True})`"
-            ) from e
-
-        if output_key not in self.node_attr_keys:
-            self.add_node_attr_key(output_key, -1)
-        elif reset:
-            self.update_node_attrs(node_ids=self.node_ids(), attrs={output_key: -1})
-
-        # node_ids are rustworkx graph ids, therefore we don't need node_id mapping
-        # and we must use RustWorkXGraph for IndexedRXGraph
-        RustWorkXGraph.update_node_attrs(
-            self,
-            node_ids=node_ids,
-            attrs={output_key: track_ids},
-        )
+            try:
+                track_node_ids, track_ids, tracks_graph = _assign_track_ids(self.rx_graph, track_id_offset)
+            except RuntimeError as e:
+                raise RuntimeError(
+                    "Are you sure this graph is a valid lineage graph?\n"
+                    "This function expects a solved graph.\n"
+                    "Often used from `graph.subgraph(edge_attr_filter={'solution': True})`"
+                ) from e
+
+            # For the IndexedRXGraph, we need to map the track_node_ids to the external node ids
+            if hasattr(self, "_map_to_external"):
+                track_node_ids = self._map_to_external(track_node_ids)  # type: ignore
+
+            # mapping to already existing track IDs as much as possible
+            if previous_id_df is not None:
+                new_id_df = pl.DataFrame({DEFAULT_ATTR_KEYS.NODE_ID: track_node_ids, output_key + "_new": track_ids})
+                merged = new_id_df.join(
+                    previous_id_df,
+                    left_on=DEFAULT_ATTR_KEYS.NODE_ID,
+                    right_on=DEFAULT_ATTR_KEYS.NODE_ID,
+                    how="left",
+                ).filter(pl.col(output_key) != -1)
+                if merged.height > 0:
+                    track_id_map = merged.unique(output_key + "_new", keep="first")
+                    track_id_map = dict(zip(track_id_map[output_key + "_new"], track_id_map[output_key], strict=True))
+                else:
```

Comment:
This is supercool.

I'm thinking of some edge cases. If there's an update where a `tracklet` 1 is split into `tracklet` 2 and 3, will this be OK? Or would this result in `track_id_map = {2: 1, and 3:1}`?

If this is the case, you may need to filter by `merged.group_by(output_key).len()` to ensure that you're not merging when remapping.

Maybe this could also happen the other way around, where a `tracklet` 2 and 3 got fused into `tracklet` 4 and it might remap back to 2 and 3.

Reply:
Hi, thank you! I'm happy to find you liked this implementation. Also thank you for finding this issue. I missed one extra `unique` function call. I believe it's fine now, but happy to have your feedback!
The later case wouldn't happen I believe (since the "4->3" row will be dropped with "4->2" existing), but I explicitly tested that [here](https://github.com/yfukai/tracksdata/blob/90fcb7d99c36bb44dbdd9788355998f79e1e0df4/src/tracksdata/graph/_test/test_graph_backends.py).